pkg/controllers/managementuser/rbac/handler_base_test.go

```diff
@@ -229,3 +229,190 @@ func Test_gatherRoles(t *testing.T) {
 		})
 	}
 }
+
+func TestCompareAndUpdateClusterRole(t *testing.T) {
+	t.Parallel()
+	numUpdatesCalled := 0
+	clusterRolesMock := &fakes2.ClusterRoleInterfaceMock{
+		UpdateFunc: func(in1 *v1.ClusterRole) (*v1.ClusterRole, error) {
+			numUpdatesCalled++
+			return &v1.ClusterRole{}, nil
+		},
+	}
+
+	tests := map[string]struct {
+		clusterRole         *v1.ClusterRole
+		roleTemplate        *v3.RoleTemplate
+		wantNumUpdatesCalls int
+	}{
+		"semantic difference": {
+			clusterRole: &v1.ClusterRole{
+				Rules: []v1.PolicyRule{
+					{
+						Verbs:     []string{"get"},
+						APIGroups: []string{""},
+						Resources: []string{"pods"},
+					},
+				},
+			},
+			roleTemplate: &v3.RoleTemplate{
+				Rules: []v1.PolicyRule{
+					{
+						Verbs:           []string{"get"},
+						APIGroups:       []string{""},
+						Resources:       []string{"pods"},
+						ResourceNames:   []string{},
+						NonResourceURLs: []string{},
+					},
+				},
+			},
+			wantNumUpdatesCalls: 0,
+		},
+		"no difference": {
+			clusterRole: &v1.ClusterRole{
+				Rules: []v1.PolicyRule{
+					{
+						Verbs:     []string{"get"},
+						APIGroups: []string{""},
+						Resources: []string{"pods"},
+					},
+				},
+			},
+			roleTemplate: &v3.RoleTemplate{
+				Rules: []v1.PolicyRule{
+					{
+						Verbs:     []string{"get"},
+						APIGroups: []string{""},
+						Resources: []string{"pods"},
+					},
+				},
+			},
+			wantNumUpdatesCalls: 0,
+		},
+		"difference": {
+			clusterRole: &v1.ClusterRole{
+				Rules: []v1.PolicyRule{
+					{
+						Verbs:     []string{"get"},
+						APIGroups: []string{""},
+						Resources: []string{"pods"},
+					},
+				},
+			},
+			roleTemplate: &v3.RoleTemplate{
+				Rules: []v1.PolicyRule{
+					{
+						Verbs:     []string{"get", "update"},
+						APIGroups: []string{""},
+						Resources: []string{"pods"},
+					},
+				},
+			},
+			wantNumUpdatesCalls: 1,
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			numUpdatesCalled = 0
+			m := manager{clusterRoles: clusterRolesMock}
+			err := m.compareAndUpdateClusterRole(test.clusterRole, test.roleTemplate)
+			assert.NoError(t, err)
+			assert.Equal(t, numUpdatesCalled, test.wantNumUpdatesCalls)
```

Comment:
Looking at the docs, it looks like these are the wrong way round?

It should be `assert.Equal(t, test.wantNumUpdatesCalls, numUpdatesCalled)`

https://pkg.go.dev/github.com/stretchr/testify/assert#Equal

Reply:
Should we be testing something more than just the number of calls? Do we care about the arguments to the calls?

Reply:
that's a good point. We can check the expected Rules are passed to the update mock. Please have a look to https://github.com/rancher/rancher/pull/44519/commits/b634a1c8600b065a1224aadfb0b2b7aa4da91899